### PR TITLE
Defaults support

### PR DESCRIPTION
**Pull request by ssenchenko**
### Description of changes
Support `Defaults` property as per design. The difference with `Globals` is that we override lists and dicts instead of merging.

The set of properties to reference in `Defaults` will be restricted to fit most common use cases:
- `Runtime` for the common section (Resolvers and Functions) only
- `InlineCode` , `CodeUri`, and `Pipeline` for Resolvers, in case customers need to override default code we generate
- `DataSource` and `DataSourceName` for Functions

We can always add more if customers ask. For now, I'm not sure that properties like `Sync`, `MaxBatchSize`, `Caching` have a lot of use. And some properties like `Id`, `Name`, `Description` don't make any sense in `Defaults` at all.

An example with `Defaults` like this
```yaml
SamGraphQL:
    Type: "AWS::Serverless::GraphQLApi"
    Propeties:
      ...
      Defaults:
        Runtime: # applied to both Resolvers and Functions
          Name: APPSYNC_JS
          RuntimeVersion: 1.0.0
        # applied to Functions only
        Functions:
          DataSource: NotesTableDataSource
      
      Resolovers:
        Query:
          getNote:
            Pipeline:
            - getNoteFunc
        Mutation:
          deleteNote:
            Pipeline:
            - deleteNoteFunc
            - notifyFunc
          saveNote:
            Pipeline:
            - saveNoteFunc
            - notifyFunc
      
      Functions:
        getNote:
          CodeUri: "function/getNote.js"
        deleteNote:
          CodeUri: "functions/deleteNote.js"
        saveNote:
          CodeUri: "functions/saveNote.js"
        notifyFunc:
          DataSource: None #override Defaults
          CodeUri: "functions/notify.js"
```

Expands to that
```yaml
SamGraphQL:
    Type: "AWS::Serverless::GraphQLApi"
    Propeties:
      ...
      Resolovers:
        Query:
          getNote:
            Runtime:
              Name: APPSYNC_JS
              RuntimeVersion: "1.0.0"
            Pipeline:
            - getNoteFunc
        Mutation:
          deleteNote:
            Runtime:
              Name: APPSYNC_JS
              RuntimeVersion: "1.0.0"
            Pipeline:
            - deleteNoteFunc
            - notifyFunc
          saveNote:
            Runtime:
              Name: APPSYNC_JS
              RuntimeVersion: "1.0.0"
            Pipeline:
            - saveNoteFunc
            - notifyFunc
      
      Functions:
        getNoteFunc:
          Runtime:
            Name: APPSYNC_JS
            RuntimeVersion: "1.0.0"
          DataSource: NotesTableDataSource
          CodeUri: "function/getNote.js"
        deleteNoteFunc:
          Runtime:
            Name: APPSYNC_JS
            RuntimeVersion: "1.0.0"
          DataSource: NotesTableDataSource
          CodeUri: "functions/deleteNote.js"
        saveNoteFunc:
          Runtime:
            Name: APPSYNC_JS
            RuntimeVersion: "1.0.0"
          DataSource: NotesTableDataSource
          CodeUri: "functions/saveNote.js"
        notifyFunc:
          Runtime:
            Name: APPSYNC_JS
            RuntimeVersion: "1.0.0"
          DataSource: None
          CodeUri: "functions/notify.js"
```

Also I had to do a couple of bug fixes like `MaxBatchSize` property forgotten for Resolver and still dangling `DataSource` for Resolvers.

### Description of how you validated changes

*Why a transformation test `*generate_code*` deleted?*  There is no need in it, Resolver generated code is present in every test with Resolvers

*Why no new `errors_` tests?*  No specific `Defaults` related errors. The only error possible is using a field that's not supported by `Defaults` but it's covered by schema.

### Checklist

- [x] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [x] Add/update [transform tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions)
    - [x] Using correct values
    - [ ] Using wrong values
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)

### Examples?

Please reach out in the comments if you want to add an example. Examples will be 
added to `sam init` through [aws/aws-sam-cli-app-templates](https://github.com/aws/aws-sam-cli-app-templates).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samtranslator/internal/schema_source/aws_serverless_graphqlapi.py [hunk 1]
```diff
@@ -152,7 +152,19 @@ class Cache(BaseModel):
...
 
+class Defaults(BaseModel):
+    # properties common for Resolvers and Functions
+    Runtime: Optional[Runtime]
+    MaxBatchSize: Optional[PassThroughProp]
+    Sync: Optional[Sync]
+    # only Resolver properties
+    Resolvers: Optional[Resolver]
+    # only Function properties
+    Functions: Optional[Function]
```
Comment (hoffa) [inline samtranslator/internal/schema_source/aws_serverless_graphqlapi.py:163]: What does it mean to have a default function?

Comment (ssenchenko) [reply samtranslator/internal/schema_source/aws_serverless_graphqlapi.py:163]: I means, we can reference here all Function properties

File: samtranslator/internal/schema_source/aws_serverless_graphqlapi.py [hunk 2]
```diff
@@ -152,7 +152,19 @@ class Cache(BaseModel):
     TransitEncryptionEnabled: Optional[PassThroughProp]
 
 
+class Defaults(BaseModel):
+    # properties common for Resolvers and Functions
+    Runtime: Optional[Runtime]
+    MaxBatchSize: Optional[PassThroughProp]
+    Sync: Optional[Sync]
```
Comment (aaythapa) [inline samtranslator/internal/schema_source/aws_serverless_graphqlapi.py:159]: Could there be a case where customers would want the defaults to only apply to Resolvers or only to Functions? What would they do in that case?

File: samtranslator/model/graphql/graphql.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    ):
+        self.resolver = resolver
+        self.function = function
+
+    @classmethod
+    def from_sam_graphql(cls, model: Properties) -> "Defaults":
+        """Create Defaults from SAM GraphQLApi model.
+
+        SAM GraphQLApi object has optional Defaults property which allows to set defaults for all
+        Functions, Resolvers, or some properties like Runtime for both Functioins and Resolvers.
```
Comment (aaythapa) [inline samtranslator/model/graphql/graphql.py:29]: nit: `Functioins` typo in comments

File: samtranslator/model/graphql/graphql.py
Comment (aaythapa) [inline samtranslator/model/graphql/graphql.py:1]: Can the `Defaults` be handled in the `sam_resources.py` file or is there a good reason it needs its own file?

File: samtranslator/internal/schema_source/aws_serverless_graphqlapi.py (hunk 1 above)
Comment (ssenchenko) [reply samtranslator/internal/schema_source/aws_serverless_graphqlapi.py:163]: Why not? Even properties like `CodeUri` or `InlineCode` have a use case to be declared in `Defaults`. If a customer uses Lambda Resolvers mostly, they might want to override the code we provide by default, but they still might want to have the same code for all Resolvers and Functions.

File: samtranslator/internal/schema_source/aws_serverless_graphqlapi.py (hunk 2 above)
Comment (ssenchenko) [reply samtranslator/internal/schema_source/aws_serverless_graphqlapi.py:159]: Yes, they will put those properties under `Resolvers` or `Functions` property:
```yaml
Defaults:
  MaxBatchSize: 10 # that's for both Resolvers and Functions
  Resolvers:
    MaxBatchSize: 20 # that's Resolvers only, it overrides '10' from above
  Functions:
    Sync:
      # this Sync setting is only for functions
```

File: samtranslator/model/graphql/graphql.py
Comment (ssenchenko) [reply samtranslator/model/graphql/graphql.py:1]: IMO a couple of reasons:
1. `sam_resources.py` is already pretty long and inconvenient because of its size
2. Having `Defaults` as class seems cleaner than function + NamedTuple (or simple Tuple or other data structure)
3. Having a class which is not a SAM resource in `sam_resources.py` doesn't sound right, even setting file size aside
4. Having `Defaults` as a nested class in `sam_resources.py` leads to extra verbosity and extra ugliness partially because we still support Python 3.7
5. Having a separate package with helper functions and data structures is a practice for other resources
6. I'd move more stuff from GraphQLApi resource to that separate graphql package, and hopefully, one day, I'll get it done :)

File: tests/translator/input/graphqlapi_resolver_ref_function_by_logical_and_function_id.yaml [hunk 4]
```diff
@@ -28,7 +28,6 @@ Resources:
           saveNote:
             InlineCode: blahblahblah
             FieldName: my_field
-            DataSource: NONE
```
Comment (hoffa) [inline tests/translator/input/graphqlapi_resolver_ref_function_by_logical_and_function_id.yaml:31]: Why was this removed?

File: samtranslator/internal/schema_source/aws_serverless_graphqlapi.py [hunk 5]
```diff
@@ -152,7 +152,19 @@ class Cache(BaseModel):
...
+    # only Resolver properties
+    Resolvers: Optional[Resolver]
```
Comment (hoffa) [inline samtranslator/internal/schema_source/aws_serverless_graphqlapi.py:161]: nit: technically these aren't correct, as this means if `Resolvers` is provided every property required by `Resolver` must be specified. It's more of a "`Resolver` but all its properties are optional", similar to TypeScript's [`Partial`](https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype), but not sure if there's anything similar

File: samtranslator/internal/schema_source/aws_serverless_graphqlapi.py (hunk 2 above)
Comment (hoffa) [inline samtranslator/internal/schema_source/aws_serverless_graphqlapi.py:159]: Is it guaranteed `Sync` will always be exactly the same for resolvers and functions?

Comment (aaythapa) [reply samtranslator/internal/schema_source/aws_serverless_graphqlapi.py:159]: Ah that makes sense, so whatever they define in the `Resolvers` will take precedence

File: samtranslator/model/graphql/graphql.py
Comment (aaythapa) [reply samtranslator/model/graphql/graphql.py:1]: Thanks for the clarification it makes sense to me. I'd say yea to moving more of the GraphQLApi resource to the graphql package because my only issue is that `Default` is the only thing thats separated from `sam_resources.py`

**Review by aaythapa — APPROVED** [review]

File: samtranslator/internal/schema_source/aws_serverless_graphqlapi.py (hunk 2 above)
Comment (ssenchenko) [reply samtranslator/internal/schema_source/aws_serverless_graphqlapi.py]: There will be no Sync in new version

File: samtranslator/internal/schema_source/aws_serverless_graphqlapi.py (hunk 5 above)
Comment (ssenchenko) [reply samtranslator/internal/schema_source/aws_serverless_graphqlapi.py]: That's a good catch. Though all props are optional now in Resolver, I thought at some point to have a separate type just in case. Well, it's a separate type in a new version anyway

File: tests/translator/input/graphqlapi_resolver_ref_function_by_logical_and_function_id.yaml (hunk 4 above)
Comment (ssenchenko) [reply tests/translator/input/graphqlapi_resolver_ref_function_by_logical_and_function_id.yaml:31]: ah, no DataSources in Resolver. They where there to be applied to all Functions in the resolver, but now we have `Defaults`

File: samtranslator/internal/schema_source/aws_serverless_graphqlapi.py (hunk 1 above)
Comment (ssenchenko) [reply samtranslator/internal/schema_source/aws_serverless_graphqlapi.py]: Restricting properties as discussed

**Review by hoffa — APPROVED** [review]
